### PR TITLE
[JAX] Add sqrtsoftplus router score function

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

Expose sqrtsoftplus routing to TE/JAX

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Expose sqrtsoftplus router enum and add corresponding tests

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_distributed_router.py [hunk 1]
```diff
@@ -160,21 +160,23 @@ def target_fwd(x):
...
+            grad_weights = jnp.linspace(0.5, 1.5, num_experts, dtype=jnp.float32)[None, :]
+
             def target_loss(x):
                 p, _ = fused_topk_with_score_function(
                     x,
                     topk=topk,
                     score_function=score_function,
                 )
-                return jnp.sum(p)
+                return jnp.sum(p * grad_weights)
```
Comment (tdophung) [inline tests/jax/test_distributed_router.py]: not sure why we are adding this?

File: tests/jax/test_distributed_router.py [hunk 2]
```diff
@@ -160,21 +160,23 @@ def target_fwd(x):
...
 
             def ref_chunk_loss(x_chunk):
-                p, _ = reference_topk_softmax_sigmoid(
+                p, _ = reference_topk_with_score_function(
                     x_chunk,
                     topk=topk,
                     score_function=score_function,
                 )
-                return jnp.sum(p)
+                return jnp.sum(p * grad_weights)
```
Comment (tdophung) [inline tests/jax/test_distributed_router.py:175]: same here

File: tests/jax/test_fused_router.py [hunk 3]
```diff
@@ -408,6 +416,55 @@ def test_topk_sigmoid(
...
 
+@pytest_parametrize_wrapper("dtype", DTYPES)
+@pytest_parametrize_wrapper(
+    "num_tokens,num_experts,topk",
+    TOPK_CASES,
+)
+@pytest_parametrize_wrapper("group_topk", GROUP_TOPK_OPTIONS)
+@pytest_parametrize_wrapper("scaling_factor", SCALING_FACTOR_OPTIONS)
+@pytest_parametrize_wrapper("enable_bias", ENABLE_BIAS_OPTIONS)
+@pytest.mark.triton
```
Comment (tdophung) [inline tests/jax/test_fused_router.py]: This is not just in your PR but I saw a lot of mark.triton on this file and it actually does not use triton here (fused router is a CUDA kernel not triton kernel so I can make anoother PR to remove all the wrong marks)

File: tests/jax/test_fused_router.py [hunk 4]
```diff
@@ -408,6 +416,55 @@ def test_topk_sigmoid(
...
+            fused_topk_with_score_function,
+            topk=4,
+            score_function=ScoreFunction.SQRTSOFTPLUS,
+        )
+    )
+
+    string_probs, string_routing_map = string_fn(logits)
+    enum_probs, enum_routing_map = enum_fn(logits)
+    assert jnp.array_equal(string_probs, enum_probs)
+    assert jnp.array_equal(string_routing_map, enum_routing_map)
```
Comment (tdophung) [inline tests/jax/test_fused_router.py]: This does not seem to be needed to test between enum and string to be passed into score_function

File: tests/jax/test_fused_router.py [hunk 5]
```diff
@@ -408,6 +416,55 @@ def test_topk_sigmoid(
...
+    run_topk_comparison(
+        dtype=dtype,
+        num_tokens=num_tokens,
+        num_experts=num_experts,
+        topk=topk,
+        use_pre_softmax=False,
+        num_groups=num_groups,
+        group_topk=group_topk,
+        scaling_factor=scaling_factor,
+        score_function="sqrtsoftplus",
```
Comment (tdophung) [inline tests/jax/test_fused_router.py]: we can also just pytest parameterize the score_function, what do you think?

**Review by tdophung — CHANGES_REQUESTED** [review]

File: tests/jax/test_fused_router.py (hunk 5 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_fused_router.py]: Good point, switched to parameterized version with a skip for enable_bias=True and softmax score function so the same test configs are still maintained

File: tests/jax/test_fused_router.py (hunk 4 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_fused_router.py]: Agreed, thanks. I've removed it now

File: tests/jax/test_fused_router.py (hunk 3 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_fused_router.py]: Removed all instances of `pytest.mark.triton` from the `test_fused_router.py` and `test_distributed_router.py` files

File: tests/jax/test_distributed_router.py (hunk 1 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_distributed_router.py]: This was added to ensure we're testing the backward pass with varied inputs. Before this change, we had varied forward `logits` input. But we were just calling `jax.grad` on this function for the backward

Since the output is an unweighted sum, the incoming gradient will be all 1s. So we're only testing the backward with constant 1 input grad

This weighting is mocking a more realistic scenario where the following layer has non-constant incoming gradients in the backward pass. I'd be okay with using randomly initialized values here too instead of linspace

This is an issue in our other tests that use `mean` or `sum` too, but for this PR I'm fixing only these tests.

Comment (tdophung) [reply tests/jax/test_distributed_router.py]: I see. I think think randomized input might be better then

Comment (jberchtold-nvidia) [reply tests/jax/test_distributed_router.py]: Switched to random in the latest commit

**Review by sylvesterkaczmarek — COMMENTED** [review]
The refactor removes the `USE_PRE_SOFTMAX_OPTIONS` parametrisation: the new `test_topk` hardcodes `use_pre_softmax=False`, and `test_te_ep_moe` also only covers the default false path. JAX `use_pre_softmax=True` therefore loses its standalone fused-router regression coverage in this PR. Restore a softmax-only `use_pre_softmax` parametrisation while adding the new score function.

**Review by tdophung — DISMISSED** [review]
LGTM pending CI

**Comment by jberchtold-nvidia** [comment]
> The refactor removes the `USE_PRE_SOFTMAX_OPTIONS` parametrisation: the new `test_topk` hardcodes `use_pre_softmax=False`, and `test_te_ep_moe` also only covers the default false path. JAX `use_pre_softmax=True` therefore loses its standalone fused-router regression coverage in this PR. Restore a softmax-only `use_pre_softmax` parametrisation while adding the new score function.

Thanks Sylvester! You're right this test coverage was missing. Fixed in the latest version

**Review by tdophung — APPROVED** [review]
